packages/scorchwrap/README.md

```diff
@@ -0,0 +1,145 @@
+
+aa[AA identifiers in policy] --> p[policy processing]
+wp[paths of the bundle] --> waa[paths mapped to AAs]
+waa --> p
+p --> w[replace AAs with numbers from a sequence]
+w --> b[use short identifiers in the bundle and inlined policy]
+
+```
+
+Webpack generates this:
+```
+const nodeCrypto = __webpack_require__(/*! crypto */ "?0b7d");
+```
+and it shows up where I list modules, but it doesn't have an associated path and I can't seem to figure out how it should be handled. It's the closest thing to a builtin in the browser. Maybe instead of handling builtins as packages we need special logic for stuff like `buffer` and `events` and this thing. Or maybe just this thing.
+
+## Manual testing
+
+run `yarn` and `yarn lib:ses` in the packages/scorchwrap folder before you begin
+
+`cd app`  
+run `npm ci`  
+run `npm test` to trigger the build  
+open dist/index.html in the browser and look at the console
+
+## Testing TODO
+
+List of what to cover ordered by priority
+
+- [ ] e2e test creating a bundle and run it to check for errors
+- [ ] cover all module types
+- [ ] configure bundle splitting
+- [ ] cover all types of runtimeRequirements with examples (stuff passed to closures along with modue and exports and `__webpack_require__`, also all the methods on `__webpack_require__`)
+- [ ] cover webpack magic comments
+
+## Features
+
+### policy
+
+There's two ways we could handle policy
+1. inline into each module at compilation time 
+   We're working per-module, so this would produce a lot of duplication but eliminate the need to store entire policy with keys for each package.
+2. add to runtime, keyed  
+   For now holding on to the entire policy seems like a better idea, we could compress the keys at compile time easily.
+   
+Current implementation inlines the policy as-is into the bundle.
+
+### package identification
+If we run a'a' on the dependencies, we're going to need means to look them up by path. It'd be nice to collect the paths first and only create IDs based on the paths included in the bundle instead of going through entire node_modules. We could collect the IDs on an earlier phase, before generate, to create a mapping from paths to IDs only for the packages involved. 
+
+After a mapping is built, the actual values of identifiers are no longer important, for bundling policy into the runtime and wrapping packages they just need to match, so it makes sense to replace them with random values before they're rendered into the bundle.
+
+
+
+
+### modes
+
+- default 
+- paranoid - adds extra checks
+
+Things it could check in paranoid mode
+- use other hooks to check if the number of bundle entries equals the number of entries processed by the wrapper
+- add a local reference before the with blocks and a function within to look it up and alert if it can be found - should trigger if a transform or plugin action breaks the withproxies. This would be best implemented by the plugin injecting a tiny module with the source to do this.
```

Comment:
is it possible to tell webpack to always run this plugin "last"?

Reply:
there's no such thing as "last"" in plugins. Loaders can have order, but what we're doing happens after loaders. For plugins it's a matter of choosing the right hooks and that's about 50% of the work I did here.